src/Data/Unfoldable1.ss

```diff
@@ -0,0 +1,25 @@
+;; -*- mode: scheme -*-
+
+(library (Data.Unfoldable1 foreign)
+  (export unfoldr1ArrayImpl)
+  (import (only (rnrs base) define lambda if set!)
+          (prefix (purs runtime srfi :214) srfi:214:)
+          (only (rnrs) do))
+
+  (define unfoldr1ArrayImpl
+    (lambda (isNothing)
+      (lambda (fromJust)
+        (lambda (fst)
+          (lambda (snd)
+            (lambda (f)
+              (lambda (b)
+                (define result (srfi:214:make-flexvector 0))
+                (define value (f b))
+                (define done #f)
+                (do () (done)
+                  (srfi:214:flexvector-add-back! result (fst value))
+                  (if (isNothing (snd value))
+                    (set! done #t)
+                    (set! value (f (fromJust (snd value))))))
+                result)))))))
```

Comment:
I would slightly prefer a recursive loop and use `let` instead of `define`:

```suggestion
  (define unfoldr1ArrayImpl
    (lambda (isNothing)
      (lambda (fromJust)
        (lambda (fst)
          (lambda (snd)
            (lambda (f)
              (lambda (b)
                (let ([result (srfi:214:make-flexvector 0)])
                  (let recur ([value (f b)])
                    (srfi:214:flexvector-add-back! result (fst value))
                    (let ([maybe (snd value)])
                      (if (isNothing maybe)
                        result
                        (recur (fromJust maybe)))))))))))))
```

Reply:
For `let` — yes, totally. As for the loop — I've heard before that Idris compiler benefits from being written [in Scheme] in very imperative style, so I assumed that's what we pick. I will fix it then.

Reply:
The tail-recursive loop will compile to a loop, so it should be fine. We do still use mutation here so that is probably the best thing performance-wise.